### PR TITLE
X509_STORE_CTX accessors.

### PR DESCRIPTION
**Pull request by richsalz**
Add some functions that were missing when a number of X509
objects became opaque (thanks, Roumen!)

### CODE REVIEW

File: crypto/x509/x509_lu.c [hunk 1]
```diff
@@ -294,25 +294,22 @@ X509_LOOKUP *X509_STORE_add_lookup(X509_STORE *v, X509_LOOKUP_METHOD *m)
     }
 }
 
-X509_OBJECT *X509_STORE_get_X509_by_subject(X509_STORE_CTX *vs, int type,
-                                                  X509_NAME *name)
+X509_OBJECT *X509_STORE_CTX_get_obj_by_subject(X509_STORE_CTX *vs, int type,
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Please change the prefix from `X509_STORE_CTX_` to `X509_STORE_`.  You're not manipulating the ctx, nor are you retrieving values from it, you'r just using it to get information from the associated store, right?

File: crypto/x509/x509_lu.c [hunk 2]
```diff
@@ -294,25 +294,22 @@ X509_LOOKUP *X509_STORE_add_lookup(X509_STORE *v, X509_LOOKUP_METHOD *m)
...
+    if (!X509_STORE_CTX_get_by_subject(vs, type, name, ret)) {
+        X509_OBJECT_free(ret);
         return NULL;
     }
     return ret;
 }
 
-int X509_STORE_get_by_subject(X509_STORE_CTX *vs, X509_LOOKUP_TYPE type,
-                              X509_NAME *name, X509_OBJECT *ret)
+int X509_STORE_CTX_get_by_subject(X509_STORE_CTX *vs, X509_LOOKUP_TYPE type,
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Same thing here, please change the prefix back from `X509_STORE_CTX_` to `X509_STORE_`.

**Comment by t-j-h** [comment]
If the functions take an X509_STORE_CTX as the first argument then they really should actually be named with X509_STORE_CTX as the prefix - so I guess @levitte and I have a different view on this one.

File: crypto/x509/x509_lu.c [hunk 3]
```diff
@@ -525,27 +526,28 @@ STACK_OF(X509_OBJECT) *X509_STORE_get0_objects(X509_STORE *v)
     return v->objs;
 }
 
-STACK_OF(X509) *X509_STORE_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
+STACK_OF(X509) *X509_STORE_CTX_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Same thing here

File: crypto/x509/x509_lu.c [hunk 4]
```diff
@@ -570,22 +572,22 @@ STACK_OF(X509) *X509_STORE_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
 
 }
 
-STACK_OF(X509_CRL) *X509_STORE_get1_crls(X509_STORE_CTX *ctx, X509_NAME *nm)
+STACK_OF(X509_CRL) *X509_STORE_CTX_get1_crls(X509_STORE_CTX *ctx, X509_NAME *nm)
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Same thing here

File: crypto/x509/x509_lu.c [hunk 5]
```diff
@@ -651,32 +653,36 @@ X509_OBJECT *X509_OBJECT_retrieve_match(STACK_OF(X509_OBJECT) *h,
 int X509_STORE_CTX_get1_issuer(X509 **issuer, X509_STORE_CTX *ctx, X509 *x)
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Ok, now I'm not so sure any more...

**Comment by levitte** [comment]
I've love to hear the reasoning for having `X509_STORE_CTX_` as a prefix just because you *use* a context.  I compare this to the *use* of `BN_CTX` in the bignum unit, we have a large number of functions that are *using* such a context, but you don't see them having `BN_CTX_` for a prefix...  only those actually manipulating the context itself (allocator, deallocator, start and end) have that prefix.

File: crypto/x509/x509_lu.c (hunk 1 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: but the arg is a store_ctx and so it's an operation on the STORE_CTX.  the old name exists in the compatibility section.

File: crypto/x509/x509_lu.c (hunk 3 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: same thing above.

**Comment by richsalz** [comment]
I think BN API's that don't take a BN are misnamed :)

**Comment by richsalz** [comment]
ping @levitte and @t-j-h .  Roumen (original person who noticed the issue) gave it his approval.

**Comment by richsalz** [comment]
And I could not find an ```BN_xxx``` API that took a ```BN_CTX``` as its first argument.

**Comment by levitte** [comment]
First argument?  What does *that* have to do with anything?  The difference as I see it is if the function's main task is to manipulate the context variable or merely use it.

**Comment by richsalz** [comment]
The context is opaque.  I have no idea what the function is doing :)

File: crypto/x509/x509_lu.c (hunk 5 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: Having looked around, I'm convinced this one is misnamed and should be called `X509_STORE_get1_issuer`.  If noone minds waiting a little bit, I'll ask DrH.

**Comment by levitte** [comment]
For the record, DrH agrees with @richsalz .

File: crypto/x509/x509_lu.c (hunk 1 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: (never mind)

File: crypto/x509/x509_lu.c (hunk 2 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: (never mind)

File: crypto/x509/x509_lu.c (hunk 3 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: (never mind)

File: crypto/x509/x509_lu.c (hunk 4 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: (never mind)

File: crypto/x509/x509_lu.c (hunk 5 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: (never mind)

**Comment by richsalz** [comment]
okay, so ... plus-one anyone?

File: crypto/x509/x509_lu.c [hunk 6]
```diff
@@ -392,11 +383,7 @@ int X509_STORE_add_crl(X509_STORE *ctx, X509_CRL *x)
 
     if (x == NULL)
         return 0;
-    obj = OPENSSL_malloc(sizeof(*obj));
-    if (obj == NULL) {
-        X509err(X509_F_X509_STORE_ADD_CRL, ERR_R_MALLOC_FAILURE);
-        return 0;
-    }
+    obj = X509_OBJECT_new();
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Check for NULL

File: crypto/x509/x509_lu.c [hunk 7]
```diff
@@ -525,27 +526,28 @@ STACK_OF(X509_OBJECT) *X509_STORE_get0_objects(X509_STORE *v)
...
+
     CRYPTO_THREAD_write_lock(ctx->ctx->lock);
     idx = x509_object_idx_cnt(ctx->ctx->objs, X509_LU_X509, nm, &cnt);
     if (idx < 0) {
         /*
          * Nothing found in cache: do lookup to possibly add new objects to
          * cache
          */
-        X509_OBJECT xobj;
+        X509_OBJECT *xobj = X509_OBJECT_new();
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Shouldn't this be checked for NULL?

File: crypto/x509/x509_lu.c [hunk 8]
```diff
@@ -570,22 +572,22 @@ STACK_OF(X509) *X509_STORE_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
...
-STACK_OF(X509_CRL) *X509_STORE_get1_crls(X509_STORE_CTX *ctx, X509_NAME *nm)
+STACK_OF(X509_CRL) *X509_STORE_CTX_get1_crls(X509_STORE_CTX *ctx, X509_NAME *nm)
 {
     int i, idx, cnt;
-    STACK_OF(X509_CRL) *sk;
+    STACK_OF(X509_CRL) *sk = sk_X509_CRL_new_null();
     X509_CRL *x;
-    X509_OBJECT *obj, xobj;
-    sk = sk_X509_CRL_new_null();
+    X509_OBJECT *obj, *xobj = X509_OBJECT_new();
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: `xobj` should be checked for NULL, no?

**Comment by levitte** [comment]
Last remarks, this is not entirely plusone-ready

File: crypto/x509/x509_lu.c (hunk 7 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: yes, fixed.

File: crypto/x509/x509_lu.c (hunk 8 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: fixed.

**Comment by richsalz** [comment]
pushed new commit to address your feedback, @levitte

File: crypto/x509/x509_lu.c [hunk 9]
```diff
@@ -359,11 +353,9 @@ int X509_STORE_add_cert(X509_STORE *ctx, X509 *x)
 
     if (x == NULL)
         return 0;
-    obj = OPENSSL_malloc(sizeof(*obj));
-    if (obj == NULL) {
-        X509err(X509_F_X509_STORE_ADD_CERT, ERR_R_MALLOC_FAILURE);
+    obj = X509_OBJECT_new();
+    if (obj == NULL)
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: So, sometimes this functions adds an error code on error, and sometimes not?  That's not consistent.

File: crypto/x509/x509_lu.c [hunk 10]
```diff
@@ -392,11 +383,9 @@ int X509_STORE_add_crl(X509_STORE *ctx, X509_CRL *x)
 
     if (x == NULL)
         return 0;
-    obj = OPENSSL_malloc(sizeof(*obj));
-    if (obj == NULL) {
-        X509err(X509_F_X509_STORE_ADD_CRL, ERR_R_MALLOC_FAILURE);
+    obj = X509_OBJECT_new();
+    if (obj == NULL)
         return 0;
-    }
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Same here, inconsistent setting of error code

File: crypto/x509/x509_lu.c [hunk 11]
```diff
@@ -524,27 +527,33 @@ STACK_OF(X509_OBJECT) *X509_STORE_get0_objects(X509_STORE *v)
     return v->objs;
 }
 
-STACK_OF(X509) *X509_STORE_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
+STACK_OF(X509) *X509_STORE_CTX_get1_certs(X509_STORE_CTX *ctx, X509_NAME *nm)
 {
     int i, idx, cnt;
-    STACK_OF(X509) *sk;
+    STACK_OF(X509) *sk = sk_X509_new_null();
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: You know, considering how it's used, I would initialise `sk` to `NULL` here...

Comment (levitte) [inline crypto/x509/x509_lu.c]: ... and do the allocation just before the `for` loop

Comment (levitte) [inline crypto/x509/x509_lu.c]: Your call

File: crypto/x509/x509_lu.c [hunk 12]
```diff
@@ -650,32 +657,36 @@ X509_OBJECT *X509_OBJECT_retrieve_match(STACK_OF(X509_OBJECT) *h,
 int X509_STORE_CTX_get1_issuer(X509 **issuer, X509_STORE_CTX *ctx, X509 *x)
 {
     X509_NAME *xn;
-    X509_OBJECT obj, *pobj;
+    X509_OBJECT *obj = X509_OBJECT_new(), *pobj = NULL;
     int i, ok, idx, ret;
 
+    if (obj == NULL)
+        return -1;
```
Comment (levitte) [inline crypto/x509/x509_lu.c]: Don't forget an error code

File: crypto/x509/x509_lu.c (hunk 9 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: The X509_STORE_new sets an error code.

File: crypto/x509/x509_lu.c (hunk 11 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: done.  pushing new commit.

File: crypto/x509/x509_lu.c (hunk 12 above)
Comment (richsalz) [inline crypto/x509/x509_lu.c]: Set by x509_object_new

File: crypto/x509/x509_lu.c (hunk 9 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: Oh!  Ok.

File: crypto/x509/x509_lu.c (hunk 11 above)
Comment (levitte) [inline crypto/x509/x509_lu.c]: Have you pushed yet?  Can't see the change...

Comment (richsalz) [inline crypto/x509/x509_lu.c]: Ha!  I did "push gitlab" not "github" :)

**Comment by levitte** [comment]
Awright, I'm satisfied.

**Comment by levitte** [comment]
I suggest waiting for Travis and Appveyor before merging

**Comment by levitte** [comment]
Travis seems happy

**Comment by richsalz** [comment]
done.

**Pull request closed without merging**